File: src/main/java/io/github/mdsimmo/bomberman/messaging/Message.java
package io.github.mdsimmo.bomberman.messaging;

import org.bukkit.ChatColor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;

import javax.annotation.CheckReturnValue;
import javax.annotation.Nonnull;
import java.util.*;

@CheckReturnValue
public final class Message implements Formattable {

    public static Message of(String text) {
        return new Message(new StringNode(text));
    }

    public static Message of(int num) {
        return of(String.valueOf(num));
    }

    private static final Message emptyMessage = Message.of("");
    public static Message empty() {
        return emptyMessage;
    }

    public static Message title(Message text, Message subtitle, int fadein, int duration, int fadeout) {
        return new Message(new TitleNode(text, subtitle, fadein, duration, fadeout));
    }

    private static Message rawMessage = new Message(new RawNode());
    public static Message rawFlag() {
        return rawMessage;
    }

    public static Message error(String s) {
        return Message.of(s).color(ChatColor.RED);
    }

    private static class Style {
        final ChatColor color;
        final Set<ChatColor> formats;

        Style(ChatColor color, Set<ChatColor> formats) {
            this.color = color;
            this.formats = formats;
        }
    }

    private static class Cursor {
        final StringBuilder content = new StringBuilder();
        final Deque<Style> colorStack = new ArrayDeque<>() {{ add(new Style(ChatColor.RESET, Set.of())); }};

        void addStyle(ChatColor color) {
            // Get what the current style is (should always be one on the stack)
            Style currentStyle = colorStack.getLast();

            // Determine what the color change will make the new style be
            Style newStyle;
            if (color.isColor()) {
                newStyle = new Style(color, currentStyle.formats);
            } else if (color.isFormat()) {
                // LinkedHashSet is used here to ensure consistent order for unit tests
                Set<ChatColor> newFormats = new LinkedHashSet<>(currentStyle.formats);
                newFormats.add(color);
                newStyle = new Style(currentStyle.color, newFormats);
            } else /* color == RESET */ {
                newStyle = new Style(ChatColor.RESET, Set.of());
            }

            // Change to the new style
            colorStack.addLast(newStyle);
            appendConversionString(currentStyle, newStyle);
        }

        private void appendConversionString(Style from, Style to) {
            // Minecraft ChatColor rules:
            //  * Adding a color/reset format will remove all previous formats
            //  * Adding a format will keep existing formats and colors

            // LinkedHashSet is used here to ensure consistent order for unit tests
            Set<ChatColor> formatsRemoved = new LinkedHashSet<>(from.formats);
            formatsRemoved.removeAll(to.formats);

            Set<ChatColor> formatsAdded = new LinkedHashSet<>(to.formats);
            formatsAdded.removeAll(from.formats);

            if (from.color != to.color || !formatsRemoved.isEmpty()) {
                // A complete reapplication is needed
                content.append(to.color);
                to.formats.forEach(content::append);
            } else if (!formatsAdded.isEmpty()) {
                // Can just apply the new formats directly
                formatsAdded.forEach(content::append);
            }
        }

        void write(String text) {
            content.append(text);
        }

        void popColor() {
            Style from = colorStack.removeLast();
            Style to = colorStack.getLast();
            appendConversionString(from, to);
        }

        @Override
        public String toString() {
            return content.toString();
        }
    }

    static class Title {
        final String title, subtitle;
        final int fadeIn, stay, fadeOut;

        Title(String title, String subtitle, int fadeIn, int stay, int fadeOut) {
            this.title = title;
            this.subtitle = subtitle;
            this.fadeIn = fadeIn;
            this.stay = stay;
            this.fadeOut = fadeOut;
        }
    }

    private interface TreeNode {
        void expand(Cursor cursor);

        // TODO handling of {raw} and {title} is very messy
        boolean isRaw();
        Optional<Title> expandTitle();
    }

    private static class StringNode implements TreeNode {
        final String text;

        private StringNode(String text) {
            this.text = text;
        }

        @Override
        public void expand(Cursor cursor) {
            cursor.write(text);
        }

        @Override
        public boolean isRaw() {
            return false;
        }

        @Override
        public Optional<Title> expandTitle() {
            return Optional.empty();
        }
    }

    private static class Joined implements TreeNode {
        final List<TreeNode> parts;

        private Joined(List<TreeNode> parts) {
            this.parts = new ArrayList<>();
            // Flatten multiple joined nodes
            for (TreeNode part : parts) {
                if (part instanceof Joined) {
                    this.parts.addAll(((Joined) part).parts);
                } else if (part != emptyMessage.contents) {
                    this.parts.add(part);
                }
            }
        }

        private Joined(TreeNode ... parts) {
            this(Arrays.asList(parts));
        }

        @Override
        public void expand(Cursor cursor) {
            for (TreeNode part : parts) {
                part.expand(cursor);
            }
        }

        @Override
        public boolean isRaw() {
            return parts.stream().anyMatch(TreeNode::isRaw);
        }

        @Override
        public Optional<Title> expandTitle() {
            return parts.stream()
                    .map(TreeNode::expandTitle)
                    .filter(Optional::isPresent)
                    .map(Optional::get)
                    .findAny();
        }
    }

    private static class Colored implements TreeNode {
        final TreeNode content;
        final ChatColor color;

        private Colored(TreeNode content, ChatColor color) {
            this.content = content;
            this.color = color;
        }

        @Override
        public void expand(Cursor cursor) {
            cursor.addStyle(color);
            content.expand(cursor);
            cursor.popColor();
        }

        @Override
        public boolean isRaw() {
            return content.isRaw();
        }

        @Override
        public Optional<Title> expandTitle() {
            return Optional.empty();
        }
    }

    // TODO Raw/Title formatting is really strange
    private static class RawNode implements TreeNode {
        @Override
        public void expand(Cursor cursor) {
        }

        @Override
        public boolean isRaw() {
            return true;
        }

        @Override
        public Optional<Title> expandTitle() {
            return Optional.empty();
        }
    }

    private static class TitleNode implements TreeNode {
        private final Message title, subtitle;
        private final int fadein, stay, fadeout;

        TitleNode(Message title, Message subtitle, int fadein, int stay, int fadeout) {
            this.title = title;
            this.subtitle = subtitle;
            this.fadein = fadein;
            this.stay = stay;
            this.fadeout = fadeout;
        }

        @Override
        public void expand(Cursor cursor) {
        }

        @Override
        public boolean isRaw() {
            return false;
        }

        @Override
        public Optional<Title> expandTitle() {
            Cursor titleCursor = new Cursor();
            title.contents.expand(titleCursor);
            String titleString = titleCursor.toString();

            Cursor subtitleCursor = new Cursor();
            subtitle.contents.expand(subtitleCursor);
            String subtitleString = subtitleCursor.toString();

            return Optional.of(new Title(titleString, subtitleString, fadein, stay, fadeout));
        }
    }

    private final TreeNode contents;

    private Message(TreeNode contents) {
        this.contents = contents;
    }

    public Message color(ChatColor color) {
        return new Message(new Colored(contents, color));
    }

    public Message append(Message text) {
        return new Message(new Joined(contents, text.contents));
    }

    public Message format(@Nonnull List<Message> args) {
        return this;
    }

    public void sendTo(CommandSender sender) {
        try {
            var sendContents = contents.isRaw() ? contents : Text.MESSAGE_FORMAT
                    .with("message", this)
                    .format().contents;
            Cursor cursor = new Cursor();
            sendContents.expand(cursor);
            if (!cursor.toString().isBlank())
                sender.sendMessage(cursor.toString());

            // Handle possible title
            if (sender instanceof Player) {
                contents.expandTitle().ifPresent(title ->
                        ((Player) sender).sendTitle(title.title, title.subtitle, title.fadeIn, title.stay, title.fadeOut));
            }

        } catch (RuntimeException e) {
            sender.sendMessage(ChatColor.RED + "Message format invalid");
        }
    }

    @Override
    public String toString() {
        Cursor cursor = new Cursor();
        contents.expand(cursor);
        return cursor.toString();
    }
}

File: src/main/java/io/github/mdsimmo/bomberman/messaging/Message.kt
package io.github.mdsimmo.bomberman.messaging

import org.bukkit.ChatColor
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import java.lang.IllegalArgumentException
import java.util.*
import javax.annotation.CheckReturnValue

/**
 * A message is an immutable coloured/formatted expanded string
 */
@CheckReturnValue
class Message private constructor(private val contents: TreeNode) : Formattable {

    // TODO the internals of this class are an utter mess

    companion object {
        fun of(text: String): Message {
            return Message(StringNode(text))
        }

        fun of(num: Number): Message {
            return of(num.toString())
        }

        val empty = of("")

        fun title(text: Message, subtitle: Message, fadeIn: Int, duration: Int, fadeOut: Int): Message {
            return Message(TitleNode(text, subtitle, fadeIn, duration, fadeOut))
        }

        val rawFlag = Message(RawNode())

        fun error(s: String): Message {
            return of(s).color(ChatColor.RED)
        }

        fun lazyExpand(text: String, context: Context): Message {
            return Message(LazyNode(text, context))
        }

        fun reference(item: Formattable, context: Context): Message {
            return Message(ReferenceNode(item, context))
        }
    }

    private data class Style constructor(
            val color: ChatColor,
            val formats: Set<ChatColor>
    )

    private class Cursor {
        val content = StringBuilder()
        val colorStack: Deque<Style> = ArrayDeque<Style>().apply {
            add(Style(ChatColor.RESET, setOf()))
        }

        fun addStyle(color: ChatColor) { // Get what the current style is (should always be one on the stack)
            val currentStyle = colorStack.last
            // Determine what the color change will make the new style be
            val newStyle = when {
                color.isColor -> Style(color, currentStyle.formats)
                color.isFormat -> {
                    // LinkedHashSet is used here to ensure consistent order for unit tests
                    val newFormats = LinkedHashSet(currentStyle.formats)
                    newFormats.add(color)
                    Style(currentStyle.color, newFormats)
                }
                else -> /* color == RESET */ {
                    Style(ChatColor.RESET, setOf())
                }
            }
            // Change to the new style
            colorStack.addLast(newStyle)
            appendConversionString(currentStyle, newStyle)
        }

        private fun appendConversionString(from: Style, to: Style) { // Minecraft ChatColor rules:
            //  * Adding a color/reset format will remove all previous formats
            //  * Adding a format will keep existing formats and colors
            // LinkedHashSet is used here to ensure consistent order for unit tests
            val formatsRemoved = LinkedHashSet(from.formats)
            formatsRemoved.removeAll(to.formats)

            val formatsAdded = LinkedHashSet(to.formats)
            formatsAdded.removeAll(from.formats)

            if (from.color != to.color || formatsRemoved.isNotEmpty()) {
                // A complete reapplication is needed
                content.append(to.color)
                to.formats.forEach { obj ->
                    content.append(obj)
                }
            } else if (formatsAdded.isNotEmpty()) {
                // Can just apply the new formats directly
                formatsAdded.forEach { obj ->
                    content.append(obj)
                }
            }
        }

        fun write(text: String) {
            content.append(text)
        }

        fun popColor() {
            val from = colorStack.removeLast()
            val to = colorStack.last
            appendConversionString(from, to)
        }

        override fun toString(): String {
            return content.toString()
        }
    }

    private data class Title(
            val title: String,
            val subtitle: String,
            val fadeIn: Int,
            val stay: Int,
            val fadeOut: Int
    )

    private interface TreeNode {
        fun expand(cursor: Cursor)
        // TODO handling of {raw} and {title} is very messy
        val isRaw: Boolean

        fun expandTitle(): Title?

        fun applyModifier(arg: Message): Formattable?
    }

    private class StringNode(val text: String) : TreeNode {
        override fun expand(cursor: Cursor) {
            cursor.write(text)
        }

        override val isRaw: Boolean
            get() = false

        override fun expandTitle(): Title? {
            return null
        }

        override fun applyModifier(arg: Message): Formattable? {
            return null
        }

    }

    private class Joined private constructor(parts: List<TreeNode>) : TreeNode {
        val parts: List<TreeNode>

        constructor(a: TreeNode, b: TreeNode) : this(listOf(a, b))

        init {
            this.parts = ArrayList()
            // Flatten multiple joined nodes
            for (part in parts) {
                if (part is Joined) {
                    this.parts.addAll(part.parts)
                } else if (part !== empty.contents) {
                    this.parts.add(part)
                }
            }
        }

        override fun expand(cursor: Cursor) {
            for (part in parts) {
                part.expand(cursor)
            }
        }

        override val isRaw: Boolean
            get() = parts.any { it.isRaw }

        override fun expandTitle(): Title? {
            return parts.firstNotNullOfOrNull { it.expandTitle() }
        }

        override fun applyModifier(arg: Message): Formattable? {
            return parts.firstNotNullOfOrNull { it.applyModifier(arg) }
        }

    }

    private class Colored(val content: TreeNode, val color: ChatColor) : TreeNode {
        override fun expand(cursor: Cursor) {
            cursor.addStyle(color)
            content.expand(cursor)
            cursor.popColor()
        }

        override val isRaw: Boolean
            get() = content.isRaw

        override fun expandTitle(): Title? {
            return null
        }

        override fun applyModifier(arg: Message): Formattable? {
            return null
        }
    }

    // TODO Raw/Title formatting is really strange
    private class RawNode : TreeNode {
        override fun expand(cursor: Cursor) {}
        override val isRaw: Boolean
            get() = true

        override fun expandTitle(): Title? {
            return null
        }

        override fun applyModifier(arg: Message): Formattable? {
            return null
        }
    }

    private class TitleNode (
            private val title: Message,
            private val subtitle: Message,
            private val fadeIn: Int,
            private val stay: Int,
            private val fadeOut: Int)
        : TreeNode {

        override fun expand(cursor: Cursor) {}

        override val isRaw: Boolean
            get() = false

        override fun expandTitle(): Title {
            val titleCursor = Cursor()
            title.contents.expand(titleCursor)
            val titleString = titleCursor.toString()
            val subtitleCursor = Cursor()
            subtitle.contents.expand(subtitleCursor)
            val subtitleString = subtitleCursor.toString()
            return Title(titleString, subtitleString, fadeIn, stay, fadeOut)
        }

        override fun applyModifier(arg: Message): Formattable? {
            return null
        }
    }

    private class LazyNode (val text: String, val context: Context) : TreeNode {
        val content: TreeNode by lazy {
            Expander.expand(text, context).contents
        }

        override fun expand(cursor: Cursor) {
            content.expand(cursor)
        }

        override val isRaw: Boolean
            get() = content.isRaw

        override fun expandTitle(): Title? {
            return content.expandTitle()
        }

        override fun applyModifier(arg: Message): Formattable? {
            return content.applyModifier(arg)
        }
    }

    private class ReferenceNode (val item: Formattable, context: Context) : TreeNode {
        val content: TreeNode by lazy {
            item.format(context).contents
        }

        override fun expand(cursor: Cursor) {
            content.expand(cursor)
        }

        override val isRaw: Boolean
            get() = content.isRaw

        override fun expandTitle(): Title? {
            return content.expandTitle()
        }

        override fun applyModifier(arg: Message): Formattable {
            return item.applyModifier(arg)
        }
    }

    fun color(color: ChatColor): Message {
        return Message(Colored(contents, color))
    }

    fun append(text: Message): Message {
        return Message(Joined(contents, text.contents))
    }

    override fun format(context: Context): Message {
        return this
    }

    override fun applyModifier(arg: Message): Formattable {
        return contents.applyModifier(arg) ?: throw IllegalArgumentException("Cannot accept extra arguments")
    }

    fun sendTo(sender: CommandSender) {
        try {
            val sendContents: TreeNode = if (contents.isRaw) {
                contents
            } else {
                // Apply the wrapper format, but not to empty values
                Switch()
                    .applyModifier(this)
                    .applyModifier(empty)
                    .applyModifier(empty)
                    .applyModifier(Text.MESSAGE_FORMAT
                        .format(Context(false).plus("message", this))
                    )
                    .format(Context(false))
                    .contents
            }
            val cursor = Cursor()
            sendContents.expand(cursor)
            if (cursor.toString().isNotBlank()) {
                // filter out empty lines
                val text = cursor.toString()
                    .split('\n', '\r')
                    .filter { it.isNotBlank() }
                    .reduce{a, b -> a + "\n" + b}
                sender.sendMessage(text)
            }
            // Handle possible title
            if (sender is Player) {
                contents.expandTitle()?.also { title ->
                    sender.sendTitle(title.title, title.subtitle, title.fadeIn, title.stay, title.fadeOut)
                }
            }
        } catch (e: RuntimeException) {
            sender.sendMessage(ChatColor.RED.toString() + "Message format invalid")
        }
    }

    override fun toString(): String {
        val cursor = Cursor()
        contents.expand(cursor)
        return cursor.toString()
    }

}